### ISSUE TITLE
google_monitoring_dashboard: resource still has perma diffs

### ISSUE THREAD

**Issue by IchordeDionysos** [hashicorp/terraform-provider-google#16173]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

I hope this helps reproduce the issue and I hope we don't need to wait until the next major version again to get this fixed 😅 

This is a long-standing annoyance, polluting the changes when applying Terraform, leading to important differences needing to be noticed more easily.

### Terraform Version

v1.6.0

### Affected Resource(s)

* google_monitoring_dashboard

### Terraform Configuration Files

<details><summary>Full Example reproducing the issue:</summary>
<p>

```tf
terraform {
  required_version = ">= 1.3.0, < 2.0.0"

  required_providers {
    google = {
      source  = "hashicorp/google"
      version = "~> 5.0"
    }
  }
}

variable "project" {
  type = string
}

variable "region" {
  type = string
}

provider "google" {
  project = var.project
  region  = var.region
}

resource "google_monitoring_dashboard" "dashboard" {

  dashboard_json = <<EOF
  {
  "dashboardFilters": [],
  "displayName": "Perma diff dashboard",
  "labels": {},
  "mosaicLayout": {
    "columns": 12,
    "tiles": [
      {
        "height": 4,
        "widget": {
          "title": "SLO Compliance: 99.9% - Good/Total Ratio - Rolling 28 days",
          "xyChart": {
            "chartOptions": {
              "mode": "COLOR"
            },
            "dataSets": [
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilter": {
                    "aggregation": {
                      "perSeriesAligner": "ALIGN_NEXT_OLDER"
                    },
                    "filter": "select_slo_compliance(\"projects/${var.project}/services/run-service/serviceLevelObjectives/run-service-rolling-availability-slo\")"
                  },
                  "unitOverride": "10^2.%"
                }
              }
            ],
            "thresholds": [
              {
                "label": "",
                "targetAxis": "Y1",
                "value": 0.999
              }
            ]
          }
        },
        "width": 12
      },
      {
        "height": 4,
        "widget": {
          "title": "SLO Compliance: 95% - 90th percentile latency under 1000ms - Rolling 28 days",
          "xyChart": {
            "chartOptions": {
              "mode": "COLOR"
            },
            "dataSets": [
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilter": {
                    "aggregation": {
                      "perSeriesAligner": "ALIGN_NEXT_OLDER"
                    },
                    "filter": "select_slo_compliance(\"projects/${var.project}/services/run-service/serviceLevelObjectives/run-service-latency-slo\")"
                  },
                  "unitOverride": "10^2.%"
                }
              }
            ],
            "thresholds": [
              {
                "label": "",
                "targetAxis": "Y1",
                "value": 0.95
              }
            ]
          }
        },
        "width": 12,
        "yPos": 4
      },
      {
        "height": 4,
        "widget": {
          "title": "Requests per response code class",
          "xyChart": {
            "chartOptions": {
              "mode": "COLOR"
            },
            "dataSets": [
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "minAlignmentPeriod": "60s",
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilter": {
                    "aggregation": {
                      "alignmentPeriod": "60s",
                      "perSeriesAligner": "ALIGN_RATE"
                    },
                    "filter": "metric.type=\"loadbalancing.googleapis.com/https/request_count\" resource.type=\"https_lb_rule\" (resource.label.backend_target_name=\"api-lb-backend-run-service\")",
                    "secondaryAggregation": {
                      "alignmentPeriod": "60s",
                      "crossSeriesReducer": "REDUCE_MEAN",
                      "groupByFields": [
                        "metric.label.\"response_code_class\""
                      ],
                      "perSeriesAligner": "ALIGN_MEAN"
                    }
                  }
                }
              }
            ],
            "thresholds": [],
            "timeshiftDuration": "0s",
            "yAxis": {
              "label": "y1Axis",
              "scale": "LINEAR"
            }
          }
        },
        "width": 12,
        "yPos": 8
      },
      {
        "height": 4,
        "widget": {
          "title": "Latency 50th & 95th percentile",
          "xyChart": {
            "chartOptions": {
              "mode": "COLOR"
            },
            "dataSets": [
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "minAlignmentPeriod": "60s",
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilter": {
                    "aggregation": {
                      "alignmentPeriod": "60s",
                      "crossSeriesReducer": "REDUCE_PERCENTILE_50",
                      "groupByFields": [],
                      "perSeriesAligner": "ALIGN_DELTA"
                    },
                    "filter": "metric.type=\"loadbalancing.googleapis.com/https/total_latencies\" resource.type=\"https_lb_rule\" (resource.label.backend_target_name=\"api-lb-backend-run-service\")"
                  }
                }
              },
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "minAlignmentPeriod": "60s",
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilter": {
                    "aggregation": {
                      "alignmentPeriod": "60s",
                      "crossSeriesReducer": "REDUCE_PERCENTILE_95",
                      "groupByFields": [],
                      "perSeriesAligner": "ALIGN_DELTA"
                    },
                    "filter": "metric.type=\"loadbalancing.googleapis.com/https/total_latencies\" resource.type=\"https_lb_rule\" (resource.label.backend_target_name=\"api-lb-backend-run-service\")"
                  }
                }
              }
            ],
            "thresholds": [],
            "timeshiftDuration": "0s",
            "yAxis": {
              "label": "y1Axis",
              "scale": "LINEAR"
            }
          }
        },
        "width": 12,
        "yPos": 12
      },
      {
        "height": 4,
        "widget": {
          "title": "Cache Hit Ratio",
          "xyChart": {
            "chartOptions": {
              "mode": "COLOR"
            },
            "dataSets": [
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilterRatio": {
                    "denominator": {
                      "aggregation": {
                        "alignmentPeriod": "60s",
                        "crossSeriesReducer": "REDUCE_SUM",
                        "groupByFields": [
                          "resource.label.\"backend_target_name\""
                        ],
                        "perSeriesAligner": "ALIGN_RATE"
                      },
                      "filter": "metric.type=\"loadbalancing.googleapis.com/https/backend_request_count\" resource.type=\"https_lb_rule\" (resource.label.backend_target_name=\"api-lb-backend-run-service\")"
                    },
                    "numerator": {
                      "aggregation": {
                        "alignmentPeriod": "60s",
                        "crossSeriesReducer": "REDUCE_SUM",
                        "groupByFields": [],
                        "perSeriesAligner": "ALIGN_RATE"
                      },
                      "filter": "metric.type=\"loadbalancing.googleapis.com/https/backend_request_count\" resource.type=\"https_lb_rule\" metric.label.\"cache_result\"=\"HIT\" (resource.label.backend_target_name=\"api-lb-backend-run-service\")"
                    }
                  }
                }
              }
            ],
            "thresholds": [],
            "timeshiftDuration": "0s",
            "yAxis": {
              "label": "y1Axis",
              "scale": "LINEAR"
            }
          }
        },
        "width": 12,
        "yPos": 16
      },
      {
        "height": 7,
        "widget": {
          "logsPanel": {
            "filter": "resource.type=\"cloud_function\"\nlog_name=\"projects/${var.project}/logs/cloudaudit.googleapis.com%2Factivity\"\nprotoPayload.request.function.labels.domain=\"run-service\"",
            "resourceNames": [
              "projects/${var.project}"
            ]
          },
          "title": "API deployments"
        },
        "width": 12,
        "yPos": 20
      }
    ]
  }
}
EOF
}

```

</p>
</details> 

<details><summary>Minimal example reproducing the issue</summary>
<p>

```tf
terraform {
  required_version = ">= 1.3.0, < 2.0.0"

  required_providers {
    google = {
      source  = "hashicorp/google"
      version = "~> 5.0"
    }
  }
}

variable "project" {
  type    = string
  default = "simpleclub-stage"
}

variable "region" {
  type    = string
  default = "europe-west1"
}

provider "google" {
  project = var.project
  region  = var.region
}

resource "google_monitoring_dashboard" "dashboard" {

  dashboard_json = <<EOF
  {
  "dashboardFilters": [],
  "displayName": "Perma diff dashboard",
  "labels": {},
  "mosaicLayout": {
    "columns": 12,
    "tiles": [
      {
        "height": 4,
        "widget": {
          "title": "SLO Compliance: 99.9% - Good/Total Ratio - Rolling 28 days",
          "xyChart": {
            "chartOptions": {
              "mode": "COLOR"
            },
            "dataSets": [
              {
                "breakdowns": [],
                "dimensions": [],
                "measures": [],
                "plotType": "LINE",
                "targetAxis": "Y1",
                "timeSeriesQuery": {
                  "timeSeriesFilter": {
                    "aggregation": {
                      "perSeriesAligner": "ALIGN_NEXT_OLDER"
                    },
                    "filter": "select_slo_compliance(\"projects/${var.project}/services/run-service/serviceLevelObjectives/run-service-rolling-availability-slo\")"
                  },
                  "unitOverride": "10^2.%"
                }
              }
            ],
            "thresholds": [
              {
                "label": "",
                "targetAxis": "Y1",
                "value": 0.999
              }
            ]
          }
        },
        "width": 12
      }
    ]
  }
}
EOF
}
```

</p>
</details> 

<details><summary>Terraform output after re-planning (with perma diffs):</summary>
<p>

```tf

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # google_monitoring_dashboard.dashboard will be updated in-place
  ~ resource "google_monitoring_dashboard" "dashboard" {
      ~ dashboard_json = jsonencode(
          ~ {
              + dashboardFilters = []
              - etag             = "dashboard-etag"
              + labels           = {}
              ~ mosaicLayout     = {
                  ~ tiles   = [
                      ~ {
                          ~ widget = {
                              ~ xyChart = {
                                  ~ dataSets     = [
                                      ~ {
                                          + breakdowns      = []
                                          + dimensions      = []
                                          + measures        = []
                                            # (3 unchanged attributes hidden)
                                        },
                                    ]
                                  ~ thresholds   = [
                                      ~ {
                                          + label      = ""
                                            # (2 unchanged attributes hidden)
                                        },
                                    ]
                                    # (1 unchanged attribute hidden)
                                }
                                # (1 unchanged attribute hidden)
                            }
                            # (2 unchanged attributes hidden)
                        },
                    ]
                    # (1 unchanged attribute hidden)
                }
              - name             = "projects/projectA/dashboards/dashboard-id"
                # (1 unchanged attribute hidden)
            }
        )
        id             = "projects/projectA/dashboards/dashboard-id"
        # (1 unchanged attribute hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

</p>
</details> 

### Expected Behavior

Re-planning shows "No changes".

### Actual Behavior

Re-planning shows changes needed for Terraform.

### Steps to Reproduce

1. `terraform apply`
2. `terraform plan`

### Important Factoids

It seems like using the `mosaicLayout` or some specific fields leads to this issue, but I'm not sure if that's indeed the case.

### References

* #7242
* GoogleCloudPlatform/magic-modules#9197

**Comment by IchordeDionysos** [comment]
@c2thorn here is the dedicated issue

**Comment by IchordeDionysos** [comment]
@c2thorn did you have time to look into this?

**Comment by philipp-wahala-db** [comment]
We face the same issue. Would be really helpful to get this resolved!

**Comment by devshree-cohere** [comment]
same issue, please fix

**Comment by pjohnston** [comment]
still an issue in 1.7.0 and provider v5.18.0

**Comment by alexandra-buica** [comment]
Hello! This issue still persists - tf version 1.7.0, provider v.5.20.0
Could someone please take a look over it? Many thanks in advance!

**Comment by tm185187** [comment]
The issue is still presists.  Tested in Terraform Version 1.8.2 and  5.26.0 for google terraform provider.

**Comment by IchordeDionysos** [comment]
Given no progress have been made on this, we switched from mosaicLayout to row layout and were able to get rid of the issues.

Also removing fields that are defaults helped.

**Comment by radsec** [comment]
++ this still exists

**Comment by danielsiwiec** [comment]
still an issue with mosaic layout. tf 1.9.5, provider 6.0.1

**Comment by danielsiwiec** [comment]
I found a solution to the problem. I used to always get the dashboard JSON from the GCP UI, using the 'Copy JSON' option. Turns out this gets a bunch of garbage that throws terraform off. Instead, we do this now and the diffs aren't showing anymore:
```shell
gcloud monitoring dashboards list --format=json | jq '.[] | select(.displayName == "YOU DASHBOARD NAME") | del(.name, .etag)'
```

**Comment by skyuchukov-bp** [comment]
> I found a solution to the problem. I used to always get the dashboard JSON from the GCP UI, using the 'Copy JSON' option. Turns out this gets a bunch of garbage that throws terraform off. Instead, we do this now and the diffs aren't showing anymore:
> 
> ```shell
> gcloud monitoring dashboards list --format=json | jq '.[] | select(.displayName == "YOU DASHBOARD NAME") | del(.name, .etag)'
> ```

This fixed the problem for us too, thanks!

**Comment by cpl-alanb** [comment]
What also works:
- Download the dashboards over the GCP UI
- Run in bash: `cat "my_dashboard.json" | jq 'del(..|nulls)' > my_dashboard_cleared.json`
- Use the "cleared" version in your deployments

jq 'del(..|nulls)' removes all null and empty values which are causing the issues in the first place

**Comment by IchordeDionysos** [comment]
@cpl-alanb that's a bit difficult to achieve if you need to dynamically and conditionally construct the JSON

**Comment by laura-lindberg-sap** [comment]
Hello! Is there an official solution for this issue?

**Comment by IchordeDionysos** [comment]
@laura-lindberg-sap no there is none.
You may be lucky with using different layout types or otherwise change your dashboards, though it's quite unpredictable what works and what doesn't...

**Comment by ensonic** [comment]
`cat "my_dashboard.json" | jq 'del(..|nulls)' > my_dashboard_cleared.json` does not help here at all. In my case the delta is about the name and etag fields that I don't have in the downlaoded json, but are present online. I can't add them on the source as the values change for each target.

**Comment by antholeole** [comment]
I have a workaround that kind of works - you lose drift dectection (i.e. if someone edits the dashboard server side, you'll have to taint the resource to get it to re-apply) but the dashboard diff only shows up if you change it locally. I haven't fully explored the ins and outs of this, but it seems to be working so far.

```hcl
 resource "local_file" "__internal_diff_detection" {
   content  = jsonencode(<some dashboard>)
   filename = "<some file name>"
 }

 resource "google_monitoring_dashboard" "nginx-pods-dashboard" {
   dashboard_json = local_file.__internal_diff_detection.content
      lifecycle {
        # ignore the json changing server side.
        ignore_changes = [dashboard_json]
        replace_triggered_by = [
		  # if the file changes locally, then you should destroy and recreate the dashboard.
          local_file.__internal_diff_detection
        ]
      }
    }
```

basically, the idea is to tie the lifecycle of the dashboard to the local file. if the local file is edited, the change is applied.

this is not perfect in any way but its a good workound for our usecases as of right now.

**Comment by tomasbackman** [comment]
I worked around this (with some help from Gemini AI) with the null_resource, trigger and lifecycle. eg:

```
locals {
  ## Config
  dashboards_filename   = "${path.module}/monitoring_dashboards.json"
}

resource "google_monitoring_dashboard" "my_board" {
  dashboard_json =jsonencode(jsondecode(file(local.dashboards_filename)))

  lifecycle {
      # Always ignore in-place updates to prevent noisy diffs from GCP-side modifications.
      ignore_changes = [dashboard_json]
  
      # Force a replacement of the dashboard only when the rebuild_trigger resource changes.
      replace_triggered_by = [null_resource.rebuild_trigger]
  }
}

resource "null_resource" "rebuild_trigger" {
  # This resource's only purpose is to change when var.dashboard_rebuild_trigger changes.
  # It is then used in the google_monitoring_dashboard's `replace_triggered_by` lifecycle
  # argument to force a replacement when desired.
  triggers = {
    trigger_value = var.dashboard_rebuild_trigger
  }
}

variable "dashboard_rebuild_trigger" {
  description = <<-EOF
    A trigger to force the replacement of dashboards.
    By default, dashboard_json is ignored to prevent noisy diffs. (As gcp adds some default values, and terraform then removes them, causing a fake diff).
    To apply local changes, just give any other input than before.
    This will trigger a replacement of the dashboard resource with the updated JSON content.
  EOF
  type        = string
  default     = "any"
}
```

**Comment by c2thorn** [comment]
Moving my comment from https://github.com/GoogleCloudPlatform/magic-modules/pull/16878#pullrequestreview-4200726964:

We've been playing whack-a-mole with this resource and the perma diffs related to it. The current diff logic is hard to parse and not sufficient.

I think this resource needs a large schema change. Across the provider we've started using a better pattern by splitting intent and state fields separately. Something like `dashboard_json` (user intent) and `effective_dashboard_json` (computed / read-only) would solve the perma diffs permanently. We've used this pattern for [labels and annotations](https://googlecloudplatform.github.io/magic-modules/best-practices/labels-and-annotations/#add-labels-and-annotations-support) which have similar dually-managed scenarios.

This would be a breaking change, but we have the 8.0.0 major release coming up in August. An overhaul of the resource would make sense for that.

We'd need buy-in from the service team, and I've started the conversation in b/304483210

**Comment by IchordeDionysos** [comment]
If this can be fundamentally be fixed via this pattern, you'd certainly have the buy-in from the customer/consumer perspective even if this a one-time breaking change.

**Comment by mmtevelde** [comment]
@c2thorn I see this didn't make it into 8.0.0, right? Does this mean it won't be done until 9.0.0 at the earliest?
